### PR TITLE
Fix resolution of delegate labels

### PR DESCRIPTION
**Pull request by takikawa**
This PR has two commits for fixing bugs related to try-delegate labels.

One fixes the printing in certain cases (similar to #1609).

The other fixes the label resolution code. Example from the spec (https://github.com/WebAssembly/exception-handling/blob/master/proposals/exception-handling/Exceptions.md):

```
try $l1
  try
    call $foo
  delegate $l1  ;; (= delegate 0)
catch
  ...
catch_all
  ...
end
```

The label `$l1` should resolve to `0` in the delegate target, which is 1 less than how typical block constructs work. The current code mis-resolves this to `1`. It should also error in cases like the following:

```
try $l1
  nop
delegate $l1
```

Because it's not possible to represent the label `$l1` in the binary format, as a `try-delegate` cannot target itself.

### CODE REVIEW

File: test/roundtrip/fold-try-delegate.txt [hunk 1]
```diff
@@ -1,28 +1,31 @@
...
+  (func
+    try
       try
         nop
-      delegate 1
-      i32.const 7
+      delegate 0
+      try
+        nop
+      delegate 0
```
Comment (aheejin) [inline test/roundtrip/fold-try-delegate.txt:11]: This test has changed.. Which fix is this relevant relevant to?

**Review by aheejin — COMMENTED** [review]
Sorry for the late reply! Can we split the two fixes into two separate PRs?

**Comment by takikawa** [comment]
> Sorry for the late reply! Can we split the two fixes into two separate PRs?

No problem, I've split the printing fix out as #1685 now.

File: test/roundtrip/fold-try-delegate.txt (hunk 1 above)
Comment (takikawa) [reply test/roundtrip/fold-try-delegate.txt]: This change should now only be in the other PR #1685 for printing. The reason this test changes, btw, is just to align it with the new test `roundtrip/try-delegate.txt` in having multiple adjacent try-delegate blocks to exercise the relevant code path.

File: src/resolve-names.cc [hunk 2]
```diff
@@ -412,7 +412,21 @@ Result NameResolver::OnCatchExpr(TryExpr*, Catch* catch_) {
...
+    // Invalid label (of the immediate try-delegate block) used.
+    if (delegate_target->index() == 0) {
+      PrintError(&delegate_target->loc, "invalid label variable for try-delegate");
+      return Result::Ok;
+    }
+    // Adjust label by one, because a delegate instruction cannot target its
+    // immediately enclosing try block.
+    delegate_target->set_index(delegate_target->index() - 1);
+  }
+  // Pop the label here as a try-delegate has no `end` instruction.
```
Comment (takikawa) [inline src/resolve-names.cc:428]: Note: I added this line in the latest revision, as this was another bug in this resolution code. (it is sufficiently tested by the `dump` test case I added---without the pop the `$try1` label was off because it came after another `try-delegate`.)

**Comment by takikawa** [comment]
In order to make the two split PRs mergeable, I changed the test addition in this PR. Instead of testing in `roundtrip`, this modifies `dump/try-delegate.txt` to ensure the label gets resolved correctly.

File: src/resolve-names.cc [hunk 3]
```diff
@@ -412,7 +412,21 @@ Result NameResolver::OnCatchExpr(TryExpr*, Catch* catch_) {
...
+  PopLabel();
```
Comment (aheejin) [inline src/resolve-names.cc:429]: I might be mistaken, but if we do this `PopLabel()` at the top, wouldn't we not need all these fixup routines?

The point of the current rule is to think of the whole `try`-`delegate` like a single branch instruction, and treat its immediate argument as such.

Comment (takikawa) [reply src/resolve-names.cc:416]: Good point, I think that will work. I pushed a commit that makes this change. It also changes the error message on the failing test case a bit, but I think that should be fine.

**Review by aheejin — APPROVED** [review]
Thanks!
